cli.rs

```diff
@@ -0,0 +1,289 @@
+        #[structopt(
+            long = "fuzzer",
+            default_value = "Honggfuzz",
+            raw(possible_values = "&Fuzzer::variants()", case_insensitive = "true")
+        )]
+        fuzzer: Fuzzer,
+    },
+    /// List all available targets
+    #[structopt(name = "list-targets")]
+    ListTargets,
+}
+
+fn main() {
+    if let Err(e) = run() {
+        eprintln!("{}", e);
+    }
+}
+
+fn run() -> Result<(), Error> {
+    use Cli::*;
+    let cli = Cli::from_args();
+
+    match cli {
+        ListTargets => {
+            for target in &get_targets()? {
+                println!("{}", target);
+            }
+        }
+        Run { target, fuzzer } => {
+            use Fuzzer::*;
+            match fuzzer {
+                Afl => run_afl(&target, None)?,
+                Honggfuzz => run_honggfuzz(&target, None)?,
+                Libfuzzer => run_libfuzzer(&target, None)?,
+            }
+        }
+        Continuous {
+            filter,
+            timeout,
+            infinite,
+            fuzzer,
+        } => {
+            let run = |target: &str| -> Result<(), Error> {
+                use Fuzzer::*;
+                match fuzzer {
+                    Afl => run_afl(&target, Some(timeout))?,
+                    Honggfuzz => run_honggfuzz(&target, Some(timeout))?,
+                    Libfuzzer => run_libfuzzer(&target, Some(timeout))?,
+                }
+                Ok(())
+            };
+
+            let targets = get_targets()?;
+            let targets = targets.iter()
+                .filter(|x| filter.as_ref().map(|f| x.contains(f)).unwrap_or(true));
+
+            if infinite {
+                for target in targets.cycle() { run(target)?; }
+            } else {
+                for target in targets { run(target)?; }
+            }
+        }
+    }
+    Ok(())
+}
+
+fn common_dir() -> Result<PathBuf, Error> {
+    let p = env::var("CARGO_MANIFEST_DIR")
+        .map(From::from)
+        .or_else(|_| env::current_dir())?
+        .join("common");
+
+    Ok(p)
+}
+
+fn create_seed_dir(target: &str) -> Result<PathBuf, Error> {
+    let seed_dir = common_dir()?.join("seeds").join(&target);
+    fs::create_dir_all(&seed_dir)?;
+    Ok(seed_dir)
+}
+
+fn create_corpus_dir(base: &Path, target: &str) -> Result<PathBuf, Error> {
+    let corpus_dir = base.join(&format!("corpus-{}", target));
+    fs::create_dir_all(&corpus_dir)?;
+    Ok(corpus_dir)
+}
+
+fn get_targets() -> Result<Vec<String>, Error> {
+    let targets_rs = fs::read_to_string(&common_dir()?.join("src/lib.rs"))?;
+    let match_fuzz_fs = Regex::new(r"pub fn fuzz_(\w+)\(")?;
+    let target_names = match_fuzz_fs
+        .captures_iter(&targets_rs)
+        .map(|x| x[1].to_string());
+    Ok(target_names.collect())
+}
+
+fn run_honggfuzz(target: &str, timeout: Option<i32>) -> Result<(), Error> {
+    let fuzzer = Fuzzer::Honggfuzz;
+    write_fuzzer_target(fuzzer, target)?;
+    let dir = fuzzer.dir()?;
```

Comment:
i tried checking out your branch and running `cargo run -- run`:

```
> cargo run -- run pulldown_cmark_read
    Finished dev [unoptimized + debuginfo] target(s) in 0.91 secs
     Running `target/debug/cli run pulldown_cmark_read`
No such file or directory (os error 2)
```

one of the downsides of the `?` approach here is that it doesn't indicate what operation we tried to do when the error happened. i haven't done any investigation yet, so it's not on this line in particular, just thought i'd mention my thoughts

Reply:
did a little investigating. this is the line that was returning the error:

https://github.com/rust-fuzz/targets/pull/103/files#diff-eaa2729b20be051a8cb0a93eb8dd2d9fR262

the `bin` directory isn't being created automatically, so i had to do `mkdir fuzzer-honggfuzz/src/bin`

Reply:
also, related to my first point, i learned you can do this with `failure`:

```diff
diff --git a/cli.rs b/cli.rs
index cf48dc4..90ef062 100644
--- a/cli.rs
+++ b/cli.rs
@@ -13,7 +13,7 @@ use std::fs;
 use std::path::{Path, PathBuf};
 use std::process::Command;

-use failure::Error;
+use failure::{Error, ResultExt};
 use regex::Regex;
 use structopt::StructOpt;

@@ -259,7 +259,8 @@ fn write_fuzzer_target(fuzzer: Fuzzer, target: &str) -> Result<(), Error> {
         .write(true)
         .create(true)
         .truncate(true)
-        .open(&path)?;
+        .open(&path)
+        .context("Could not write fuzzer target")?;

     let source = template.replace("###TARGET###", &target);
     file.write_all(source.as_bytes())?;
```

not suggesting we should do that everywhere, just wanted to share

Reply:
I've just sprinkled a bunch of error context's all over the place and also added automatically creating this `*/src/bin/` directories

Reply:
Wow, I didn't about `failure::ErrorExt`, that's really neat!